### PR TITLE
Ghosts can assume control of minions

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Title. Also move taking ssd mobs form verb to an observer action
You have to pay 500 points in hive upgrade in order to allow ghost to take control of minions.

## Why It's Good For The Game

Need the minion nerf, and possibly more balance later on

## Changelog
:cl:
add: Ghosts can assume control of minions if the hive upgrade "Smart Minions" is bought. It is 500 psy points
/:cl:

### CODE REVIEW

**Comment by Mrrpip** [comment]
Beetlin Ramis.

**Comment by private-tristan** [comment]
I feel this will be wildly unbalanced, I like it.

**Comment by novaepee** [comment]
Marines get vatborn, xenomorphs get minions. Good game design, ngl.

**Review by Kuro020 — APPROVED** [review]

**Comment by Rayfod** [comment]
The AI of the mantis sometimes permanently takes over while you are controlling the mob.
The spit of the scorpion does also sometimes get stuck, leaving you spitting at whatever you targeted at that time, until you hit the spit button again.
I do not know about the bettle, but I would assume that there are a few bugs for them as well.

**Comment by Luxyboo** [comment]
i have an issue with this because of the balance issues, currently scorpion and beetle are on par, if not better, than spitter and defender respectively. possibly nerfing the damage, speed, or ability to heal would be some concepts to take into account when it comes to balancing these player controlled ai. maybe minions only on low pop but on high pop they arent available.

**Comment by SrSC0ndition** [comment]
Scorpion being a better spitter? Beetle over defender? Have you read these lines without a smile? If so, I am afraid you know little bit nothing about spitter and defender
The only issue with controllable minions being they have no cooldown on taking, only limitation being the amount of spawners and silos, other round xenos had cheesy 7 minions on top of their usual pop, and you don't want get mauled by 3 beetles at the same time

File: code/datums/actions/observer_action.dm [hunk 1]
```diff
@@ -40,3 +40,56 @@
...
+	var/mob/dead/observer/dead_owner = owner
+	if(GLOB.key_to_time_of_death[owner.key] + TIME_BEFORE_TAKING_BODY > world.time && !dead_owner.started_as_observer)
+		to_chat(owner, span_warning("You died too recently to be able to take a new mob."))
+		return
+
+	var/list/mob/living/free_ssd_mobs = list()
+	for(var/mob/living/ssd_mob AS in GLOB.ssd_living_mobs)
+		if(is_centcom_level(ssd_mob.z))
+			continue
+		if(ssd_mob.afk_status == MOB_RECENTLY_DISCONNECTED)
```
Comment (hyper2snyper) [inline code/datums/actions/observer_action.dm:58]: combine this and the above

File: code/datums/actions/observer_action.dm [hunk 2]
```diff
@@ -40,3 +40,56 @@
...
+		to_chat(owner, span_warning("That player hasn't been away long enough. Please wait [round(timeleft(new_mob.afk_timer_id) * 0.1)] second\s longer."))
+		return FALSE
+
+	if(is_banned_from(owner.ckey, new_mob?.job?.title))
+		to_chat(owner, span_warning("You are jobbaned from the [new_mob?.job.title] role."))
+		return
+	message_admins(span_adminnotice("[owner.key] took control of [new_mob.name] as [new_mob.p_they()] was ssd."))
+	log_admin("[owner.key] took control of [new_mob.name] as [new_mob.p_they()] was ssd.")
+	new_mob.transfer_mob(owner)
+	if(ishuman(new_mob))
```
Comment (hyper2snyper) [inline code/datums/actions/observer_action.dm:92]: early return for my sanity

**Review by hyper2snyper — CHANGES_REQUESTED** [review]
we can merge after

**Comment by novaepee** [comment]
I miss it when players get to control minons before hive has to buy abilities for ghosts to control minions.

**Comment by AndroidMK2** [comment]
I would apply nerfs to minions but Only to those currently being player controlled, when a ghost takes over a minion the minion itself has lower stats than the same minion without a player controlling it so that minions are still useful when there's no one to control them and not too strong when they are.

**Comment by Mrrpip** [comment]
> I would apply nerfs to minions but ONLY to those currently being player controlled, when a ghost takes over a minion the minion itself has lower stats than the same minion without a player controlling it so that minions are still useful when there's no one to control them and not too strong when they are.

But that'd leave you with a gimped minion that doesn't have perfect accuracy.

**Comment by Dueu** [comment]
I have played minions a bit and if the health reduction is in general I prefer not to play as minions, such a nerf to that degree makes minions much less useful to have, player controlled minions is not a guaranteed as the amount of players connected changes depending the time of the day. As AndroidMK2 recommended, the nerf should come if a player is controlling it.

**Comment by BraveMole** [comment]
> I have played minions a bit and if the health reduction is in general I prefer not to play as minions, such a nerf to that degree makes minions much less useful to have, player controlled minions is not something guaranteed as the amount of players connected changes depending the time of the day. As AndroidMK2 recommended, the nerf should come if a player is controlling it.

The health nerf is not that huge. Before, minions were running away at 40% health. I have removed that behaviour, and some removed 40% of beetle health (so it has no effect but costing less bullets and never escaping, but it was almost never espacting) and 25% of mantis health (this was actually a buff of controlled one). 

The only major nerf for ai minions is the speed. But i have a pr to fix that

File: code/datums/actions/observer_action.dm [hunk 3]
```diff
@@ -40,3 +40,54 @@
 		add_selected_frame()
 		return
 	remove_selected_frame()
+
+/datum/action/observer_action/take_ssd_mob
+	name = "Take ssd mob"
```
Comment (XSlayer300) [inline code/datums/actions/observer_action.dm:45]:
```suggestion
	name = "Take SSD mob"
```

File: code/datums/actions/observer_action.dm [hunk 4]
```diff
@@ -40,3 +40,54 @@
...
+		if(is_centcom_level(ssd_mob.z) || ssd_mob.afk_status == MOB_RECENTLY_DISCONNECTED)
+			continue
+		if(isxeno(ssd_mob))
+			var/mob/living/carbon/xenomorph/potential_minion = ssd_mob
+			if((potential_minion.xeno_caste.caste_flags & CASTE_IS_A_MINION) && !potential_minion.hive.upgrades_by_name[GHOSTS_CAN_TAKE_MINIONS].times_bought)
+				continue
+		free_ssd_mobs += ssd_mob
+
+	if(!free_ssd_mobs.len)
+		to_chat(owner, span_warning("There aren't any ssd mobs."))
```
Comment (XSlayer300) [inline code/datums/actions/observer_action.dm:65]:
```suggestion
		to_chat(owner, span_warning("There aren't any SSD mobs."))
```

**Pull request merged**